### PR TITLE
nixos/podman, podman: switch to `netavark` network stack

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

~~Breaking changes are restricted so we can't merge this until after branch off.~~

### CODE REVIEW

**Comment by zowoq** [comment]
cc @nixos/podman @roberth @mikroskeem

I'd like to get this merged in the next couple of weeks.

**Comment by luxus** [comment]
Result of `nixpkgs-review pr 199965` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Review by luxus — APPROVED** [review]

**Comment by mikroskeem** [comment]
Result of `nixpkgs-review pr 199965` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>podman</li>
    <li>podman-unwrapped</li>
  </ul>
</details>

Any reason why nixos-test is blacklisted?

**Comment by mikroskeem** [comment]
After resetting podman and verifying that netavark is network backend by checking `podman system info`, everything looks good so far (DNS etc.).

Though I'm thinking that we could keep network backend configurable, few deployments of mine are relying on CNI plugins which netavark does not support.

**Comment by zowoq** [comment]
> Though I'm thinking that we could keep network backend configurable.

This ha been discussed previously, see: https://github.com/NixOS/nixpkgs/pull/160410#issuecomment-1051415528.

Maintaining two implementations in tree is a hassle, the UX is bad with needing to reset podman between them and CNI config is just json. 

People can set `network.network_backend = cni` and write their own config to in `/etc` if they want to, this just means they lose a bit of sugar.

**Review by mikroskeem — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `carnix` and `cratesIO` has been removed due to being unmaintained, use alternatives such as [naersk](https://github.com/nix-community/naersk) and [crate2nix](https://github.com/kolloch/crate2nix) instead.
 
+- `podman` now uses the `netavark` network stack. Users will need to run `podman system reset --force` once after upgrading their systems, this command will delete all of your containers and images.
```
[the remark this answers is not in the document]
Comment (zowoq) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Updated.

**Comment by zowoq** [comment]
I'll merge this in 24hrs.

**Comment by PedroHLC** [comment]
In case someone is running NixOS in servers in production and needs to keep old containers running without recreating it: https://gist.github.com/PedroHLC/6634f606100296628a5adb8f79da2e01

File: nixos/modules/virtualisation/podman/default.nix [hunk 2]
```diff
@@ -164,11 +152,26 @@ in
...
+          dns_enabled = false;
+          driver = "bridge";
+          id = "0000000000000000000000000000000000000000000000000000000000000000";
+          internal = false;
+          ipam_options = { driver = "host-local"; };
+          ipv6_enabled = false;
+          name = "podman";
+          network_interface = "podman0";
+          subnets = [{ gateway = "10.88.0.1"; subnet = "10.88.0.0/16"; }];
+        } // cfg.defaultNetwork.settings);
```
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/podman/default.nix:167]: Does this even has an effect? Wouldn't the conditional in line 156 deactivate the entire file?

Comment (zowoq) [reply nixos/modules/virtualisation/podman/default.nix:167]: > Does this even has an effect?

Yes, the dns test works.

> Wouldn't the conditional in line 156 deactivate the entire file?

Yes, file isn't created by default, only if `defaultNetwork.settings` is used.

**Pull request merged**